### PR TITLE
[DSCP-171] Block retrieval logic

### PR DESCRIPTION
**Pull request by volhovm**
This PR adds simple block retrieval. Whenever a new node starts, it can catch up with the chain. Fork resolution is not supported yet.

### CODE REVIEW

File: witness/src/Dscp/Witness/Workers/Worker.hs [hunk 1]
```diff
@@ -6,72 +6,88 @@ module Dscp.Witness.Workers.Worker
...
+                             , ccallback processTipMsg
+                             , ccallback processBlocks
+                             ]
+
+    processNewBlock nId btq block = do
+        let header = bHeader block
+        tip <- runSdM getTipHeader
+        let tipHash = headerHash tip
+        if | hPrevHash header == tipHash -> applyNewBlock block
+           | hDifficulty header > hDifficulty tip -> do
```
Comment (Heimdell) [inline witness/src/Dscp/Witness/Workers/Worker.hs]: Is the "difficulty" the length of the chain?

File: core/src/Dscp/Core/Foundation/Witness.hs [hunk 2]
```diff
@@ -259,7 +259,7 @@ newtype SlotId = SlotId Word64
     deriving (Eq, Ord, Num, Show, Generic, Buildable)
 
 -- | Chain difficulty.
-newtype Difficulty = Difficulty Word64
+newtype Difficulty = Difficulty { unDifficulty :: Word64 }
```
[the remark this answers is not in the document]
Comment (volhovm) [reply core/src/Dscp/Core/Foundation/Witness.hs]: We don't have a strict policy regarding strictness annotations. My personal choice is to put them into records so one can't forget to initialise field, thus I don't use them for newtypes. We can discuss it in slack if you want to.

File: core/src/Dscp/Util.hs [hunk 3]
```diff
@@ -13,6 +13,11 @@ module Dscp.Util
        , assert
        , assertJust
 
+         -- * Maybe conversions
+       , nothingToThrow
+       , nothingToFail
+       , nothingToPanic
```
[the remark this answers is not in the document]
Comment (volhovm) [reply core/src/Dscp/Util.hs]: This is not intentional, but I like it. http://memeshappen.com/media/created/2016/11/Nothing-to-see-here-Move-along-Leah.jpg

File: witness/src/Dscp/Snowdrop/Validators.hs [hunk 4]
```diff
@@ -5,26 +5,25 @@ module Dscp.Snowdrop.Validators
...
 
 ----------------------------------------------------------------------------
 -- Validator
 ----------------------------------------------------------------------------
 
-type IOCtx chgAccum = SD.IOCtx chgAccum Ids Values
-
 validator :: SD.Validator Exceptions Ids Proofs Values (IOCtx chgAccum)
-validator = mempty
+validator = _baseValidator
```
[the remark this answers is not in the document]
Comment (volhovm) [reply witness/src/Dscp/Snowdrop/Validators.hs]: Sure. Let's wire it in a separate task -- this should be done carefully.

File: witness/src/Dscp/Witness/Workers/Worker.hs (hunk 1 above)
Comment (volhovm) [reply witness/src/Dscp/Witness/Workers/Worker.hs]: Yes, it's a common notation.

File: core/src/Dscp/Core/Foundation/Witness.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Martoon-00) [reply core/src/Dscp/Core/Foundation/Witness.hs]: How can a new type have strictness annotation :thinking_face:

[the remark this answers is not in the document]
Comment (volhovm) [reply core/src/Dscp/Core/Foundation/Witness.hs]: Oh wait, yes.

File: witness/src/Dscp/Snowdrop/Validators.hs [hunk 5]
```diff
@@ -95,12 +94,51 @@ blkStateConfig ::
...
+      bsfConfig = cfg
+    , bsfApplyPayload = \(SPayload txs _) -> do
+          chg <- either SD.throwLocalError pure $
+              foldM SD.mappendChangeSet def $ map SD.txBody txs
+          undo <- SD.liftERoComp $ SD.computeUndo chg
+          forM_ txs $ \tx -> do
+              SD.liftERoComp $ SD.runValidator vld tx
+              SD.modifyRwCompChgAccum (SD.txBody tx)
+          pure undo
+    , bsfApplyUndo = SD.modifyRwCompChgAccum
```
[the remark this answers is not in the document]
Comment (Heimdell) [reply witness/src/Dscp/Snowdrop/Validators.hs]: Ok, we'll leave it as it is now.

File: witness/src/Dscp/Snowdrop/Validators.hs [hunk 6]
```diff
@@ -95,12 +94,51 @@ blkStateConfig ::
...
+    , bsfStoreBlund = \blund -> do
+          let blockRef = SD.bcBlockRef cfg (SD.blkHeader $ SD.buBlock blund)
+          let chg = SD.ChangeSet $ M.singleton (SD.inj $ SD.BlockRef blockRef)
+                                               (SD.New $ SD.inj blund)
+          SD.modifyRwCompChgAccum chg
```
[the remark this answers is not in the document]
Comment (Heimdell) [reply witness/src/Dscp/Snowdrop/Validators.hs]: Same as above, lets leave this as it is.

**Pull request merged**